Improve shell scripts

zapashcanon opened the pull request:
Hi,

I fixed some of the things spotted by ShellCheck in the shell scripts.

I used the following command: `find -O3 . -type f -name '*.sh' -exec shellcheck -x {} \;`.

When running the test suite, I got at the end:

```sh
TEST      coq-makefile/quick2vo
make[3] : on entre dans le répertoire « /home/zapashcanon/Programmation/OCaml/coq/test-suite »
BUILDING SUMMARY FILE
FAILURES
0m01.19s  | Compilers/Z/Bounds/Pipeline                   0m01.19s  | Compilers/Z/Bounds/Pipeline                    | 0m01.04s  || +0m00.14s | +14.42%     coq-makefile/plugin2/run.sh...Error!
    coq-makefile/arg/run.sh...Error!
    coq-makefile/compat-subdirs/run.sh...Error!
    coq-makefile/coqdoc1/run.sh...Error!
    coq-makefile/coqdoc2/run.sh...Error!
    coq-makefile/emptyprefix/run.sh...Error!
    coq-makefile/extend-subdirs/run.sh...Error!
    coq-makefile/findlib-package/run.sh...Error!
    coq-makefile/latex1/run.sh...Error!
    coq-makefile/merlin1/run.sh...Error!
    coq-makefile/mlpack1/run.sh...Error!
    coq-makefile/mlpack2/run.sh...Error!
    coq-makefile/multiroot/run.sh...Error!
    coq-makefile/only/run.sh...Error!
    coq-makefile/plugin1/run.sh...Error!
    coq-makefile/plugin3/run.sh...Error!
    coq-makefile/quick2vo/run.sh...Error!
    coq-makefile/timing/run.sh...Error!
    coq-makefile/uninstall1/run.sh...Error!
    coq-makefile/uninstall2/run.sh...Error!
    coq-makefile/validate1/run.sh...Error!
    coq-makefile/vio2vo/run.sh...Error!
make[3]: *** [Makefile:192: report] Error 1
make[3] : on quitte le répertoire « /home/zapashcanon/Programmation/OCaml/coq/test-suite »
make[2]: *** [Makefile:110: all] Error 2
make[2] : on quitte le répertoire « /home/zapashcanon/Programmation/OCaml/coq/test-suite »
make[1]: *** [Makefile.build:556: test-suite] Error 2
make[1] : on quitte le répertoire « /home/zapashcanon/Programmation/OCaml/coq »
make: *** [Makefile:188: submake] Error 2
```

But I'm not sure it's because of the changes I made...

SkySkimmer commented on dev/build/windows/configure_profile.sh:
```diff
@@ -17,27 +17,30 @@ donefile=~/.bash_profile.upated
+        echo export http_proxy="http://$1"
+        echo export https_proxy="http://$1"
+        echo export ftp_proxy="http://$1"
+      } >> $rcfile
```
AFAICT all echoing is redirected to rcfile so you could as well do `exec >> $rcfile` at the beginning.

zapashcanon replied:
@SkySkimmer: done, there's already a `echo >> $rcfile` of which I don't understand the purpose.

SkySkimmer commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -1246,21 +1246,22 @@ function get_cygwin_mingw_sources {
       local ARCHIVEESC=${ARCHIVE//+/\\+}
-      local SOURCE=`egrep -A 1 "install: ($CYGWINARCH|noarch)/release/[-+_/a-z0-9]*$ARCHIVEESC" $TARBALLS/setup.ini | tail -1 | cut -d " " -f 2`
+      SOURCE=$(grep -E -A 1 "install: ($CYGWINARCH|noarch)/release/[-+_/a-z0-9]*$ARCHIVEESC" $TARBALLS/setup.ini | tail -1 | cut -d " " -f 2)
+      local SOURCE
```
Shouldn't `local` come first?

SkySkimmer commented on dev/ci/ci-sf.sh:
```diff
@@ -1,12 +1,12 @@
-wget -qO- ${sf_lf_CI_TARURL}  | tar xvz
-wget -qO- ${sf_plf_CI_TARURL} | tar xvz
-wget -qO- ${sf_vfa_CI_TARURL} | tar xvz
+mkdir -p "${CI_BUILD_DIR}" && cd "${CI_BUILD_DIR}" || exit
```
This `|| exit` shouldn't be necessary considering we just did `mkdir -p`.
If you still want to silence shellcheck it should be `exit 1`.

SkySkimmer commented on test-suite/coq-makefile/coqdoc1/run.sh:
```diff
@@ -9,7 +9,13 @@ make html mlihtml
+  while IFS= read -r -d '' d
+  do
+    pushd "$d" >/dev/null && find . && popd >/dev/null
+  done <   <(find tmp -name user-contrib)
```
Shouldn't you use `find -print0` if you use `read -d ''`? Same for other copies of this pattern.

SkySkimmer commented on test-suite/coq-makefile/template/init.sh:
```diff
@@ -1,10 +1,11 @@
 find . -maxdepth 1 -not -name . -not -name _test -exec cp -r '{}' -t _test ';'
 
-cd _test
+cd _test || exit
```
`exit 1`

SkySkimmer commented on test-suite/coq-makefile/timing/run.sh:
```diff
@@ -73,12 +73,12 @@ cat A.v.timing.diff
 for ext in "" .desired; do
-    for file in A.v.timing.diff; do
-        cat ${file}${ext} | sed s"/${INFINITY}/${INFINITY_REPLACEMENT}/g" | sed s":|\s*N/A\s*$:| ${INFINITY_REPLACEMENT}:g" | sed s'/[0-9]*\.[0-9]*//g' | sed s'/0//g' | sed s'/  */ /g' | sed s'/ *$//g' | sed s'/+/-/g' | sort > ${file}${ext}.processed
+    for file in A.v.timing.diff/*; do
```
This change is not correct, do `file=A.v.timing.diff` removing the loop instead.

SkySkimmer commented on test-suite/coq-makefile/timing/run.sh:
```diff
@@ -73,12 +73,12 @@ cat A.v.timing.diff
 done
-for file in A.v.timing.diff; do
-    diff -u $file.desired.processed $file.processed || exit $?
+for file in A.v.timing.diff/*; do
```
Incorrect, see previous comment.

SkySkimmer commented on test-suite/misc/exitstatus.sh:
```diff
@@ -1,7 +1,8 @@
-printf "On ill-typed input, coqc returned $P.\n"
-if [ $N = 1 -a $P = 1 ]; then exit 0; else exit 1; fi
+printf "On ill-typed input, coqtop returned %s.\n" "$N"
+printf "On ill-typed input, coqc returned %s.\n" "P"
```
Missing $ on P.
I don't know why we use printf here, echo should be fine too.

gares commented:
> @gares what's with the find tmp -name user-contrib in various coq-makefile tests? I don't see how the tmp directory gets created.

I see many `make install DSTROOT="$PWD/tmp"` that will create tmp

SkySkimmer commented:
OK

Gaëtan Gilbert

On 22/02/2018 14:13, Enrico wrote:
>     @gares <https://github.com/gares> what's with the find tmp -name
>     user-contrib in various coq-makefile tests? I don't see how the tmp
>     directory gets created.
> 
> I see many |make install DSTROOT="$PWD/tmp"| that will create tmp
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coq/coq/pull/6809#issuecomment-367676509>, or mute 
> the thread 
> <https://github.com/notifications/unsubscribe-auth/ACWQ7FdIpvqysjsNqcSV9pXnAHU4Hle8ks5tXWfogaJpZM4SOa_2>.
>

SkySkimmer commented on test-suite/coq-makefile/timing/run.sh:
```diff
@@ -40,7 +40,7 @@ INFINITY_REPLACEMENT="+.%" # assume that if the before time is zero, we expected
+        grep -v 'warning: undefined variable' < ${file}${ext} | sed s"/${INFINITY}/${INFINITY_REPLACEMENT}/g" | sed s'/[0-9]//g' | sed s'/ *$//g' | sed s":|\s*N/A\s*$:| ${INFINITY_REPLACEMENT}:g" | sed s'/^-*$/------/g' | sed s'/  */ /g' | sed s'/\(Total.*\)-\(.*\)-/\1+\2+/g' > ${file}${ext}.processed
     done
 done
 for file in time-of-build-before.log time-of-build-after.log time-of-build-both.log; do
```
It may make sense as well to change the loops here to be more like
```bash
for file in ...; do
  for ext in "" .desired; do
    grep ...
  done
  echo "cat $file"
  ...
done

zapashcanon replied:
Will do. Did you see my comment on your first review of this loop ?

SkySkimmer replied:
Yes, the above was meant as an answer yo it.

SkySkimmer commented on dev/build/windows/configure_profile.sh:
```diff
@@ -14,30 +14,31 @@
+
 if [ ! -f $donefile ] ; then
 
     echo >> $rcfile
```
This redirection can be removed (this line just adds a newline to rcfile).

zapashcanon replied:
Done. I added a `touch $rcfile` at the beginning.

zapashcanon commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -903,10 +903,10 @@ function make_camlp5 {
     # Somehow my virus scanner has the boot.new/SAVED directory locked after the move for a second => repeat until success
     sed -i 's/mv boot.new boot/until mv boot.new boot; do sleep 1; done/' Makefile
-    log1 make world.opt $MAKE_OPT
+    log1 make world.opt "$MAKE_OPT"
```
@SkySkimmer, I suspect this to be incorrect

zapashcanon commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -1136,16 +1136,16 @@ function make_coq {
       log1 make
     else
-      make $MAKE_OPT
+      make "$MAKE_OPT"
```
TODO

zapashcanon commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -1187,8 +1187,8 @@ function make_mingw_make {
-    logn configure ./configure --build=$BUILD --host=$HOST --target=$TARGET --prefix="$PREFIXCOQ" --program-prefix=$TARGET-
-    log1 make $MAKE_OPT
+    logn configure ./configure --build="$BUILD" --host="$HOST" --target="$TARGET" --prefix="$PREFIXCOQ" --program-prefix="$TARGET-"
+    log1 make "$MAKE_OPT"
```
TODO

zapashcanon commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -1213,13 +1213,13 @@ function make_gcc {
         # --disable-decimal-float seems to be required
         # --with-sysroot="$PREFIX"  results in configure error that this is not an absolute path
-    log1 make $MAKE_OPT
+    log1 make "$MAKE_OPT"
```
TODO

SkySkimmer commented:
@zapashcanon this PR is waiting for you, will you have time to work on it in the foreseeable future? Otherwise we should either close it or have someone else take over.

zapashcanon commented:
Yes, will update it this week-end.

zapashcanon commented:
@SkySkimmer, I rebased and fixed the `$MAKE_OPT` thing in `dev/build/windows/makecoq_mingw.sh`.

The two things left to do are:
- to wait for the build, see if the `$MAKE_OPT` fix fixed everything on the mingw build
- find what causes the build to failed on my machine (and probably in the CI too, I don't remember), but I don't have time for this today, will try to have a look soon

Zimmi48 commented:
The whole CI passed except for Windows packaging scripts but they have started failing in all builds since 2 days ago. Usually, we would ignore this failure and merge anyways but because this pull request is changing the packaging scripts too, we can't do it in this case.

You are not responsible for the review requests: it is a GitHub functionality that we just activated.

@MSoegtropIMC is absent from the code owner file but should probably review this PR (and BTW Michael, the failures of the packaging jobs that we have right now on AppVeyor are yet again due to a download from Camlcity having failed: this is a very unreliable site, see #6520).

Zimmi48 commented on dev/ci/user-overlays/00669-maximedenes-ssr-merge.sh:
```diff
@@ -1,3 +1,4 @@
+#!/bin/sh
```
I suggest we do not add these shebangs to the files in `user-overlays` as they are not supposed to be executed on their own but sourced somewhere else.

zapashcanon replied:
Actually, the shebang was added so Shellcheck can check the script is correct according to what is available in a POSIX shell. The shebang won't change anything when sourcing the file, so it's harmless. Moreover, if someone wants to edit the script, it's better to have it as you know it was written as a POSIX shell, which could explain some things in the script. And I don't think someone will try to execute this on its own, as there's no reason to do it.

Anyway, if you still want me to remove all these shebangs, will do. :)

Zimmi48 replied:
It's either remove them or update the README in the same directory.

zapashcanon replied:
Updated the `README`.

Zimmi48 commented:
I've relaunched the Windows build.

@zapashcanon 
> find what causes the build to failed on my machine (and probably in the CI too, I don't remember), but I don't have time for this today, will try to have a look soon

Does it still fail on your machine?

@SkySkimmer: is your request for changes still relevant?

@gares: do you want to have a look at the changes to the coq_makefile test-suite?

Who's self-assigning this PR?

MSoegtropIMC commented on dev/build/windows/configure_profile.sh:
```diff
@@ -14,30 +14,29 @@
 rcfile=~/.bash_profile
 donefile=~/.bash_profile.upated
 
+exec >> $rcfile
```
Using exec redirection: technically this is fine, but I think this construct is not so well known, so I don't know if this change is good from a understandability / maintainability point of view. My guess is less than 10% of people with good bash experience understand this.

zapashcanon replied:
Added a comment with a link to understand `exec >>`.

MSoegtropIMC commented on dev/ci/ci-common.sh:
```diff
@@ -20,7 +20,8 @@ else
         export CI_BRANCH="$CIRCLE_BRANCH"
     else # assume local
-        export CI_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
+        CI_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
```
This change looks a bit arbitrary. All other cases are export var=value. Why is it split here?

SkySkimmer replied:
This way if git fails the command fails, otherwise export will hide the failure.

MSoegtropIMC commented on dev/ci/ci-common.sh:
```diff
@@ -35,10 +36,10 @@ ls "$COQBIN"
 
 # shellcheck source=ci-basic-overlay.sh
-source "${ci_dir}/ci-basic-overlay.sh"
+. "${ci_dir}/ci-basic-overlay.sh"
```
Replacing source with . : technically fine, but reusable shell scripts should be optimized for readability. I use . interactively and source in multi-use scripts, and I think many follow this approach.

Applies to all other identical changes

zapashcanon replied:
Actually, I tend to replace `source` by `.` because `.` is posix and `source` is not. Here we're using bash, so `source` is OK but still... I usually prefer posix things when possible. I think `.` is well known and so it seems fine to me to use it. Anyway, if you still want me to change it, will do.

MSoegtropIMC replied:
Interesting. I used source since ever also on non Linux / GNU based Unixes (IRIX, AIX, HP-UX) but I can confirm that it is not Posix, so I agree that . is preferable.

MSoegtropIMC commented on test-suite/coq-makefile/coqdoc1/run.sh:
```diff
@@ -9,7 +9,13 @@ make html mlihtml
+  while IFS= read -r -d '' d
+  do
+    pushd "$d" >/dev/null && find . && popd >/dev/null
+  done < <(find tmp -name user-contrib -print0)
```
I think (my personal opinion) non posix constructs like <() deserve a small comment like 

for <(cmd) see https://www.gnu.org/software/bash/manual/html_node/Process-Substitution.html)

It is tricky to google something like <() and it is not that widely used. Code should be optimized for readability by average experienced engineers.

MSoegtropIMC replied:
P.s.: Inst't there a more concise way to write this and still handle paths with spaces in it properly?

zapashcanon replied:
Added a comment. I have no better idea at the moment...

MSoegtropIMC commented on test-suite/coq-makefile/findlib-package/run.sh:
```diff
@@ -7,7 +7,8 @@ export OCAMLPATH=$OCAMLPATH:$PWD/findlib
   # the only way I found to pass OCAMLPATH on win is to have it contain
   # only one entry
-  export OCAMLPATH=`cygpath -w $PWD/findlib`
+  OCAMLPATH=$(cygpath -w "$PWD"/findlib)
```
As above, technicall fine, but as above the export split is not consistent (elsewhere export var=value is used)

zapashcanon replied:
As it was said above, we must separate assigning and exporting when the value comes from a command, so that, the script will fail if the command fails.

MSoegtropIMC commented on test-suite/coq-makefile/timing/precomputed-time-tests/run.sh:
```diff
@@ -4,7 +4,8 @@ set -x
+export COQLIB
 
 ./001-correct-diff-sorting-order/run.sh || exit $?
 ./002-single-file-sorting/run.sh || exit $?
```
Nothing you changed, but isn't || exit $? redundant with set -e?

zapashcanon replied:
That's true. Will open an issue on Shellcheck.

zapashcanon replied:
Fixed. Opened [an issue on ShellCheck](https://github.com/koalaman/shellcheck/issues/1163).

SkySkimmer commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -383,7 +383,7 @@ function build_conf_make_inst {
   # Check if any files matching the pattern exist
   if [ "$(echo $1)" != "$1" ] ; then
-    install -D -t $2 $1
+    install -D -t "$2" "$1"
```
This is incorrect and causes a failure as `$1` is a glob.

SkySkimmer replied:
Also it may be better to rewrite this so there's less messing with globs.
For instance 
```bash
function install_glob {
    if [ -e "$2" ]; then
        install -D -t "$@"
    fi
}
```
This involves a change in calling convention:
```bash
install_glob "$target" "$foo"*"bar" # note glob not quoted and argument order changed
```

MSoegtropIMC replied:
I agree that it would be more elegant to switch the arguments and do the glob expansion in the call to install_glob.
Also if someone has an elegant way to handle null globs in bash - please educate me. Preferably the test inside install_glob should be on the argument count then.

SkySkimmer replied:
> Also if someone has an elegant way to handle null globs in bash - please educate me. Preferably the test inside install_glob should be on the argument count then.

Do we set the bash option for expanding null globs to an empty array? If yes then we can test argument count, otherwise it expands to itself (ie `*` => `*`) and we have to test whether that's a valid file.

MSoegtropIMC replied:
nullglob is not set in my scripts - a lot of commands do unintended things if this is set , e.g. a grep bla *.txt would read from stdin instead of failing or returning nothing. And afaik bash has no way to do it just in one case.

MSoegtropIMC replied:
Maybe we should use find instead of glob - it is a bit more to write in the calls to install_glob, but I think it is a more maintainable solution.

SkySkimmer replied:
I think globs are fine for this.

zapashcanon replied:
Fixed.

gares commented:
[troll]: ban SpellCheck

zapashcanon commented:
I fixed everything that seems relevant to me.

Currently, the tests are still failing on my machine.

I just noticed there's still a lot of warnings raised by ShellCheck (I though I fixed all of them, but I was wrong), I will fix them but in a separate commit to ease the review process.

Would you be interested in adding ShellCheck to the CI ? If yes, I can do it in this PR and also add shellcheck directives in comments to prevent raising falses positives.

Zimmi48 commented on dev/ci/user-overlays/README.md:
```diff
@@ -14,3 +14,5 @@ fi
 
 (`CI_PULL_REQUEST` and `CI_BRANCH` are set in [`ci-common.sh`](/dev/ci/ci-common.sh))
+
+Even if the shell scripts in this directory contain a shebang, they should not be directly executed.
```
That's not what I meant by "update the README". I don't think this sentence is necessary: it is not dangerous to directly execute this shell scripts, it will just do nothing.
What I meant by "update the README" is to add the shebang in the quoted shell above: it says what `00669-maximedenes-ssr-merge.sh` contains but you've modified it.
BTW these overlay scripts are not meant to be kept (except this example one) so instead of patching them all, you could also simply remove them in a preliminary commit.

zapashcanon replied:
Done.

Zimmi48 commented:
I added the "needs: fixing" label to reflect the failing tests on your machine.

Yes, I think that adding SpellCheck to CI would be relevant: like test cases for fixed bugs, like the linter for whitespaces, once we start fixing things that we consider bad, we should add tests to ensure that they don't come back.

MSoegtropIMC commented:
What exactly is failing? In case it is in my area (Windows build scripts) I can have a look.

One more note: It should also be checked if the resulting Windows installer installs a reasonable Coq. During development I tested this by comparing an installed Coq with a built Coq. If things go wrong during install file selection, the scripts might succeed to create an installer, but the result might not be useful.
Which brings up the question: should we have a test in CI which actually runs the installer and checks if the installed Coq matches expectations?

Zimmi48 commented:
> should we have a test in CI which actually runs the installer and checks if the installed Coq matches expectations?

Yes, absolutely! Ideally, we should install it then run the test-suite using the installed version. However, we risk being too close from timeout to do so on AppVeyor. Maybe it will be simpler when we eventually move to GitLab CI.

SkySkimmer commented:
>we should install it then run the test-suite using the installed version

That's what gitlab does.

SkySkimmer commented on dev/build/windows/makecoq_mingw.sh:
```diff
@@ -383,7 +383,7 @@ function build_conf_make_inst {
   # Check if any files matching the pattern exist
   if [ "$(echo $1)" != "$1" ] ; then
-    install -D -t $2 $1
+    install -D -t $2 "$1"
```
This is still wrong, please look at my previous comment again.

zapashcanon replied:
Fixed.

SkySkimmer commented:
I'm not convinced that being strict on shellcheck compliance is worth doing. I'm also worried about new shellcheck versions recognising new potential issues which means new errors would appear out of thin air.

MSoegtropIMC commented:
@Zimmi48 : just running Coq tests is not sufficient, and likely also not efficient. Things like CoqIDE also need to be checked. I think it is best to compare the installation destination folder with the folder from which the installer picks it up with some rules.

@SkySkimmer : what I mean is:
- build Coq
- test Coq
- build the Windows installer
- install Coq using this installer
- somehow test the installation

Are the last 2 steps already implemented somewhere.

SkySkimmer commented:
Gitlab does:
- build coq
- `make install`
- move to a new container keeping only the install directory (+ a couple misc files)
- run the test suite with the installed Coq

MSoegtropIMC commented:
*   build coq
  *   make install
  *   move to a new container keeping only the install directory (+ a couple misc files)
  *   run the test suite with the installed Coq

On Windows things are a bit more complicated. What is created by make install is taken as a template for the NSIS Windows installer, which is part of the official Coq deployment (downloadable on the Coq web site). What I meant is run the NSIS Windows installer and check if the second install folder created by it matches the folder structure created by make install (with folder names patched in a few files).
Intel Deutschland GmbH
Registered Address: Am Campeon 10-12, 85579 Neubiberg, Germany
Tel: +49 89 99 8853-0, www.intel.de
Managing Directors: Christin Eisenschmid, Christian Lamprechter
Chairperson of the Supervisory Board: Nicole Lau
Registered Office: Munich
Commercial Register: Amtsgericht Muenchen HRB 186928

ejgallego commented:
If you are curious, for those building Coq with Dune, everything is run from the "installed" version, there is no notion of having a "local" build anymore, this is IMHO pretty good.

MSoegtropIMC commented:
@ejgallego : my Windows build script can also create either a Windows installer or install Coq right away with make install. But for building plugins the local build and the cgywin based OCaml cygwin to MinGW cross tollchain is still required. I am still working on supplying an OCaml suitable for building plugins which works without cygwin or a bash, so that the local build is no longer required for building plugins.

ejgallego commented:
> @ejgallego : my Windows build script can also create either a Windows installer or install Coq right away with make install. But for building plugins the local build and the cgywin based OCaml cygwin to MinGW cross tollchain is still required. I am still working on supplying an OCaml suitable for building plugins which works without cygwin or a bash, so that the local build is no longer required for building plugins.

The point with Dune is that `install` doesn't exists anymore [technically that's not exactly true but you get the idea].

Note that it is not the case that you should need a local build to build plugins since at least 8.7.2, you can build plugins and toplevels using standard `ocamlfind` predicates from an installed Coq version that is properly registered with `ocamlfind`. Unfortunately `coq_makefile` is not aware of this, but both `ocamlbuild` and Dune work for it.

I still fail to understand why Coq has to provide so many custom toolsets in Windows instead of tying to adapt to the OCaml standard solution for that platform.

MSoegtropIMC commented:
@ejgallego : the reason is that the majority of users of the Windows installer are students who will not write plugins. For these users one wants to have a simple "one click" install method which does not require installing OCaml first. This installer is mostly used in university courses teaching Coq.

Of cause it might be that there is a better way to create this installer such that it is more compatible with platforms which allow writing plugins or support Coq native compute. I am open to such discussions. The method supported by my build scripts doesn't seem to be widely used. But such discussions would belong in a separate thread. I would start such a discussion with a poll on who writes plugins on Windows in which way.

ejgallego commented:
> @ejgallego : the reason is that the majority of users of the Windows installer are students who will not write plugins. For these users one wants to have a simple "one click" install method which does not require installing OCaml first. This installer is mostly used in university courses teaching Coq.

Ummm, I guess that OCaml is light enough (~2 MiB) as to be included in the main package without too much trouble. Linux packages for example draw `ocaml` and `findlib` as a dependency, so you get a fully working environment. 

IMVHO the way to go here is to register the Coq development setup in the standard, findlib compatible way even on windows. Users would notice no difference AFAICT.

Zimmi48 commented:
Please keep the discussion on-topic!

zapashcanon commented:
So, tests are still failing on my machine. Here's the steps I'm following, maybe I'm doing something wrong:

```sh
make clean
./configure
# using all the defaults
make build
make test-suite
```

End output of `make test-suite`:
```sh
TEST      coq-makefile/quick2vo
make[3] : on entre dans le répertoire « /home/leo/Programmation/OCaml/coq/test-suite »
BUILDING SUMMARY FILE
FAILURES
0m01.19s  | Compilers/Z/Bounds/Pipeline                   0m01.19s  | Compilers/Z/Bounds/Pipeline                    | 0m01.04s  || +0m00.14s | +14.42%     coq-makefile/extend-subdirs/run.sh...Error!
    coq-makefile/arg/run.sh...Error!
    coq-makefile/compat-subdirs/run.sh...Error!
    coq-makefile/coqdoc1/run.sh...Error!
    coq-makefile/coqdoc2/run.sh...Error!
    coq-makefile/emptyprefix/run.sh...Error!
    coq-makefile/findlib-package/run.sh...Error!
    coq-makefile/latex1/run.sh...Error!
    coq-makefile/merlin1/run.sh...Error!
    coq-makefile/mlpack1/run.sh...Error!
    coq-makefile/mlpack2/run.sh...Error!
    coq-makefile/multiroot/run.sh...Error!
    coq-makefile/only/run.sh...Error!
    coq-makefile/plugin1/run.sh...Error!
    coq-makefile/plugin2/run.sh...Error!
    coq-makefile/plugin3/run.sh...Error!
    coq-makefile/quick2vo/run.sh...Error!
    coq-makefile/timing/run.sh...Error!
    coq-makefile/uninstall1/run.sh...Error!
    coq-makefile/uninstall2/run.sh...Error!
    coq-makefile/validate1/run.sh...Error!
    coq-makefile/vio2vo/run.sh...Error!
make[3]: *** [Makefile:194: report] Error 1
make[3] : on quitte le répertoire « /home/leo/Programmation/OCaml/coq/test-suite »
make[2]: *** [Makefile:112: all] Error 2
make[2] : on quitte le répertoire « /home/leo/Programmation/OCaml/coq/test-suite »
make[1]: *** [Makefile.build:552: test-suite] Error 2
make[1] : on quitte le répertoire « /home/leo/Programmation/OCaml/coq »
make: *** [Makefile:190: submake] Error 2
```

ejgallego commented:
> So, tests are still failing on my machine. Here's the steps I'm following, maybe I'm doing something wrong:

Likely you are missing some dependency of the test suite. You want to run with `TRAVIS=1` set so you get more verbose output, or just run that test manually. Dependencies are documented somewhere I think, but we are doing a pretty bad job about it [makefile should check and complain about missing stuff instead of the current failure]

zapashcanon commented:
OK thanks, I thought `./configure` would spot missing dependencies for me. Will try to see what's missing.

Currently, ignoring useless ones, here are the remaining warnings:

```sh
find -O3 . -type f -name '*.sh' -exec shellcheck -e SC2034,SC2154,SC1090,SC1091  {} \;

In ./dev/macosify_accel.sh line 1:
#!/usr/bin/sed -f
^-- SC1008: This shebang was unrecognized. Note that ShellCheck only handles sh/bash/dash/ksh.


In ./dev/macosify_accel.sh line 2:
s/^;\{0,1\} *\(.*\)<Control>\(.*\)$/\1<Primary>\2/
            ^-- SC2035: Use ./*glob* or -- *glob* so names with dashes won't become options.
                                    ^-- SC1001: This \1 will be a regular '1' in this context.
                                               ^-- SC1001: This \2 will be a regular '2' in this context.


In ./dev/macosify_accel.sh line 3:
s/^;\{0,1\} *\(.*\)<Alt>\(.*\)$/\1<Control>\2/
            ^-- SC2035: Use ./*glob* or -- *glob* so names with dashes won't become options.
                                ^-- SC1001: This \1 will be a regular '1' in this context.
                                           ^-- SC1001: This \2 will be a regular '2' in this context.


In ./dev/build/windows/makecoq_mingw.sh line 226:
      if file -i $2.$3 | grep text/html; then
                 ^-- SC2086: Double quote to prevent globbing and word splitting.
                    ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./dev/build/windows/makecoq_mingw.sh line 227:
        echo Download failed: $1/$2.$3
                              ^-- SC2086: Double quote to prevent globbing and word splitting.
                                 ^-- SC2086: Double quote to prevent globbing and word splitting.
                                    ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./dev/build/windows/makecoq_mingw.sh line 229:
        cat $2.$3
            ^-- SC2086: Double quote to prevent globbing and word splitting.
               ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./dev/build/windows/makecoq_mingw.sh line 386:
  if [ "$(echo $1)" != "$1" ] ; then
        ^-- SC2116: Useless echo? Instead of 'cmd $(echo foo)', just use 'cmd foo'.
               ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./dev/build/windows/makecoq_mingw.sh line 458:
  tr '/' '\\' < "/build/filelists/$1" | sed -r 's/^\.(.*)\\([^\\]+)$/SetOutPath $INSTDIR\\\1\nFile ${COQ_SRC_PATH}\\\1\\\2/' > "/build/filelists/$1.nsh"
           ^-- SC1003: Want to escape a single quote? echo 'This is how it'''s done'.
                                               ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.


In ./dev/build/windows/makecoq_mingw.sh line 1350:
    echo 'COQ_SRC_SUBDIRS:=$(filter-out plugins/%,$(COQ_SRC_SUBDIRS)) plugins/syntax' >> Makefile.coq.local
         ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.
```

SkySkimmer commented on dev/tools/merge-pr.sh:
```diff
@@ -78,8 +78,8 @@ if [ -z "$REMOTE" ]; then
 REMOTE_URL=$(git remote get-url "$REMOTE" --push)
-if [ "$REMOTE_URL" != "$OFFICIAL_REMOTE_URL" -a \
-     "$REMOTE_URL" != "$OFFICIAL_REMOTE_URL.git" ]; then
+if [ "$REMOTE_URL" != "$OFFICIAL_REMOTE_URL" ] & \
```
Should be &&

zapashcanon replied:
Oops. Fixed.

SkySkimmer commented:
>macosify_accel.sh

As indicated by the shebang this is a sed script not a shell script (unless you call sed a shell I guess). Probably should be renamed.

>In ./dev/build/windows/makecoq_mingw.sh line 226, 227, 229

Just quote these.

>rest

Ignore those.

SkySkimmer commented on dev/ci/ci-iris-lambda-rust.sh:
```diff
@@ -18,6 +18,7 @@ git_checkout "${lambdaRust_CI_BRANCH}" "${lambdaRust_CI_GITURL}" "${lambdaRust_C
 Iris_VERSION=$(grep -F coq-iris < "${lambdaRust_CI_DIR}/opam" | grep -E 'dev\.([0-9.-]+)' -o)
 Iris_URL=$(opam show "coq-iris.$Iris_VERSION" -f upstream-url)
+# shellcheck disable=SC2046
 read -r -a Iris_URL_PARTS <<< $(echo "$Iris_URL" | tr '#' ' ')
```
Why not quote here?

zapashcanon replied:
Fixed.

ejgallego commented:
> I thought ./configure would spot missing dependencies for me.

There are already some open bugs in this front, but please open one with severity high if that happens to you. I think configure is intended to check only for coq-build deps, so it should be the test-suite makefile who checks.

zapashcanon commented:
OK, so I think I fixed everything which was asked, opened an issue about the missing dependencies not being spotted. Some warnings remains, let's just ignore them as @SkySkimmer said. Is there anything left to do ?

MSoegtropIMC commented:
What about the failing appveyor opam build? It says it can't find bogomips for running complexity tests - shouldn't be that hard to fix.

And how about the SF test? Is it normal that it fails? I know SF quite well and if there are portability issues, I could look into them.

I didn't follow all discussions on CI, but I would generally expect that CI tests pass or are switched off in some agreed way until fixed.

SkySkimmer commented:
SF is fixed in master: #7121

SkySkimmer commented:
>What about the failing appveyor opam build? It says it can't find bogomips for running complexity tests - shouldn't be that hard to fix.

This is normal for windows and unrelated to the failure (if it can't find bogomips it skips some tests with that warning). The failure is the nondeterministic test for #3263, see #5981.

MSoegtropIMC commented:
@SkySimmer : thanks for the update on CI status ;-)

Good to go then from my point of view then.

Since this affects multiple code owners, who is supposed to merge/pull it ? @maximedenes ?

MSoegtropIMC commented:
P.S.: in the future I would appreciate pull requests, which are split up by code owner, unless there are unavoidable dependencies. It is hard to track improvement requests with so many changes.

Zimmi48 commented:
In principle, it is the code owner for the main component being affected. Here there is not really a main component but the CI and the Windows scripts being the most changed parts, it could be either @SkySkimmer (who has already supervised this PR the most) or you @MSoegtrop.

We might want to get a green light from @gares for the changes to the coq-makefile test-suite although it has no registered code owner right now.

MSoegtropIMC commented:
Sorry for such a github newbee question (we have our own git based tools here), but I really don't understand the commit dates github shows me. I wanted to download the latest version of this pull request, do a local build and see if the installer actually works as expected. What confuses me is that github shows me different dates for one commit, so that I am a bit lost to what is the commit of the initial pull request and what is the latest fixed commit.

If I open a specific file e.g.:

https://github.com/zapashcanon/coq/blob/sh/dev/build/windows/makecoq_mingw.sh

github says that commit b2075a9 is 2 days ago:

zapashcanon zapashcanon Improve shell scripts  b2075a9 2 days ago

Clicking on the short ID gives me this long id:

https://github.com/zapashcanon/coq/commit/b2075a99fa1cb2910c375f743b5501c4462e562a


When I open the folder in which this file is:

https://github.com/zapashcanon/coq/tree/sh/dev/build/windows

Github says that commit b2075a9 is from February 21st:

zapashcanon zapashcanon Improve shell scripts Latest commit b2075a9 on Feb 21

It is the same long id:

https://github.com/zapashcanon/coq/commit/b2075a99fa1cb2910c375f743b5501c4462e562a

I somehow don't understand how one commit (identified by one long id) can have two different dates.

The pull request also shows only one commit dates February 21st but @zapashcanon claims he did all requested changes.

I almost get the impression that commits are mutable in github (backed by "edit this file in the github editor" menu options). Possibly I am simply outdated, but I think this would be orthogonal to the space spanned by the fundamental ideas of git. I hope there is some other good explanation for what I see.

Anyway if someone can tell me how to create a local copy of whatever the latest version of this pull request is, I would appreciate it. I did a

git clone -b sh git@github.com:zapashcanon/coq.git coq-6809

but I am no longer sure this is really what I want ...

MSoegtropIMC commented:
Since the commit is indeed from February 21st and this is the only commit in the branch / pull request, I don't understand where the changes  mentioned by @zapashcanon are:

> OK, so I think I fixed everything which was asked, opened an issue about the missing dependencies not being spotted.

I want to test the version including the changes mentioned above.

SkySkimmer commented:
>Since the commit is indeed from February 21st 

That's incorrect, the commit was first authored 02/21 and last modified 04/01.

MSoegtropIMC commented:
> That's incorrect, the commit was first authored 02/21 and last modified 04/01.

but as far as I understand only metadata of the commit did change, things like comments and information that it is compatible (can be automatically merged) with other commits. The commit itself (identified by a specific source diff and the hash code of this diff) did not change. But @zapashcanon claimed above he did do all requested changes:

> OK, so I think I fixed everything which was asked, opened an issue about the missing dependencies not being spotted.

Where are these changes? In some place I can't find them or not yet in github?

SkySkimmer commented:
>but as far as I understand only metadata of the commit did change, things like comments and information that it is compatible (can be automatically merged) with other commits. The commit itself (identified by a specific source diff and the hash code of this diff) did not change. But @zapashcanon claimed above he did do all requested changes:

That is incorrect. The commit did change, retaining the information that it is a modified version of a commit first written in February (the original commit has been garbage collected by now except maybe on @zapashcanon's machine). If you look at the actual files you should be able to find the changes.

What happened is basically
```sh
# run this in a fresh directory and look around before and after any command where you don't understand what's going on (also read the corresponding man page to figure out what specifics to look at)
git init
echo foo > file
git add file
git commit -m initial
# commit aaaaaa
# pretend this is the initial state of Coq master when @zapashcanon started working

echo '"foo"' > file
git add file
git commit -m message1
git show --pretty=fuller
# Author(date) and Commit(date) identical = date1
# parent aaaaaa
# message is message1
# diff adds quotes in file
# pretend this is the state when the PR was new

# PR comments ask for more foo

echo morefoo >> file
git add file && git commit --amend -m message2
git show --pretty=fuller
# Authordate = date1
# Commitdate = date2
# message is message2
# diff adds quotes and morefoo in file
# this is the state now
```

> Where are these changes? In some place I can't find them or not yet in github?

In the commit.

>but as far as I understand only metadata of the commit did change

That's not how it works, commits are the sum state of all data (the files) and metadata (committer name, date, commit message, but also which commits are parents) and is identified by the hash of all that stuff. As such modifying a commit is the act of making a new commit based on it (like modifying an immutable record in ocaml `fun x -> {x with foo = bar}`).

The diff is derived information: it is the diff (according to the `--diff-algorithm` option) between the data of the first parent and the data of the commit.

SkySkimmer commented:
By the way the appveyor log should have shown the log of the failing test, what is going on here? cc @ejgallego

MSoegtropIMC commented:
@SkySkimmer : thanks for the detailed explanation!

In my work environment we don't use forks for individual developers as common in github, so a push / commit --amend / push is a no no here. But I see that with forks for each contributor a push / commit --amend / push is unproblematic because no one else can base work on a commit until the change is pulled by the common origin.

The conclusion is that commits are mutable as long as they stay on a single user fork. Thanks for making this clear! It simplifies the work a lot.

SkySkimmer commented:
> The conclusion is that commits are mutable

Maybe you mean something different from what I'm hearing but that sounds 
wrong to me. I would rather say that the history of a branch is mutable.

Gaëtan Gilbert

On 03/04/2018 13:53, MSoegtropIMC wrote:
> @SkySkimmer <https://github.com/SkySkimmer> : thanks for the detailed 
> explanation!
> 
> In my work environment we don't use forks for individual developers as 
> common in github, so a push / commit --amend / push is a no no here. But 
> I see that with forks for each contributor a push / commit --amend / 
> push is unproblematic because no one else can base work on a commit 
> until the change is pulled by the common origin.
> 
> The conclusion is that commits are mutable as long as they stay on a 
> single user fork. Thanks for making this clear! It simplifies the work a 
> lot.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coq/coq/pull/6809#issuecomment-378223537>, or mute 
> the thread 
> <https://github.com/notifications/unsubscribe-auth/ACWQ7MiJv9uYJqmup6-mc5KYIinvsV-bks5tk2KigaJpZM4SOa_2>.
>

ejgallego commented:
These "amended" commits are introduced by `git rebase`, and it usually makes sense to preserve the original date even if it is convenient to linearize history. It is good practice tho to reset the date when doing a significant `amend` IMHO.

Zimmi48 commented:
Yep, the `--ignore-date` option of `git rebase` can be used to reset the commit dates to the present time.

MSoegtropIMC commented:
Since I never used it, I must admit that I need more experience with commit --amend before I can say I fully understood what it does. I am confused because there seem to be two definitions of "a commit".

On the one hand is the list of commits for 6809 there is only one commit dated February 21st:

https://github.com/coq/coq/pull/6809/commits

As far as I understood the contents (in the sense of modified file diff) of this commit changed over time with commit --amend although it retained the ID. Or does the ID shown in github change with a commit --amend? If so, it seems to be really wrong / confusing to show the original date in the list linked above.

The metadata of "a commit" you listed above goes in the same direction: there is one initial date and a later commit date. How can one call such a thing immutable?

On the other hand the typical definition of a commit in git is a change diff together with some metadata like a comment and repo state information, identified by a hash of the diff and the metadata. This is obviously immutable, since a change would change the hash and thus the commit ID.

My understanding was that commit --amend more or less purges the old commit from history and creates a new one, but this would then have a new ID and would be a new commit. This doesn't seem to match what github shows. Or is it so that this is just metadata that git remembers that some new commit replaced a previous one? Still I find it strange that these have the same ID. Or maybe they haven't and the ID changes and just the original date is kept ...

My apologizes for getting way of topic here and digging into git basics. I hope I am not the only one having issues understanding this.

MSoegtropIMC commented:
@Zimmi48 and  @ejgallego : we are discussing here commit --amend with significant changes (in the sense of changes a reviewer wants to see / check), not changes done by a trivial rebase.

zapashcanon commented:
Actually I used `git commit --amend --no-edit` a lot of times, and `git rebase -i HEAD~2` one time to merge the two commits I made (setting the last one as a `fixup`).

SkySkimmer commented:
>Or maybe they haven't and the ID changes and just the original date is kept ...

That's it.

You may want to look at the doc for --amend: https://git-scm.com/docs/git-commit#git-commit---amend

It's difficult to find the previous hashes but we can look at previous travis builds in https://travis-ci.org/coq/coq/pull_requests (you may need to click "show more" at the bottom a few times)
So the most recent is https://travis-ci.org/coq/coq/builds/360897034, travis says it tested bb35ac2. Clicking on it reveals that this is the automatic merge of b2075a9 (current for this PR) with whatever was master at the time.
An older build is https://travis-ci.org/coq/coq/builds/360882593 which tested c6008d7, this one is a merge of e856fe9 into an older master, e856fe9 being an old version of this PR.

MSoegtropIMC commented:
@SkySkimmer : thanks again!

I also reread the definition of what a commit technically is:

https://git-scm.com/book/en/v2/Git-Internals-Git-Objects

So all fine: commits are immutable as they have always been.

Just github's display is confusing using either the date the commit was initially authored (February 1st for b2075a9) or the date the commit was done (April 1st for b2075a9) depending on the situation. I would generally prefer the latter, but github seems to prefer the first.

Zimmi48 commented:
I agree GitHub display can be very confusing. For instance, even on a linear branch (with no fork / merge points), it will show commits not in the order they are on the branch but in the order they were initially authored (so moving commits around with a `git rebase` without option `--ignore-date` can sometimes result in very confusing commit lists on GitHub).

SkySkimmer commented:
In #7102:
>OK. Anyways, it is going to conflict with the shell script PR. Is it ready to be merged? If yes, I'll rebase afterwards.

Not sure if we're waiting on review from @gares and owners @maximedenes @ppedrot.

Zimmi48 commented:
I'm not sure of what @ppedrot is an owner in this PR.

maximedenes commented:
It would be nice to extend `hub` so that it can tell what files each owner should look at for a given PR.

MSoegtropIMC commented:
Note: I checked if the windows installer built by the script still works as expected (do a local build, install and compare Coq template folder and install destination folder).

Check result is ok.

Zimmi48 commented:
@SkySkimmer What does your script show regarding what files are owned by @ppedrot in this PR?

SkySkimmer commented:
ide/MacOS/relatify_with-respect-to_.sh: @ppedrot (/ide/)

Gaëtan Gilbert

On 05/04/2018 10:43, Théo Zimmermann wrote:
> @SkySkimmer <https://github.com/SkySkimmer> What does your script show 
> regarding what files are owned by @ppedrot <https://github.com/ppedrot> 
> in this PR?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coq/coq/pull/6809#issuecomment-378863202>, or mute 
> the thread 
> <https://github.com/notifications/unsubscribe-auth/ACWQ7AOGrZrO0_S9OP661r7pCkvABgSqks5tldkzgaJpZM4SOa_2>.
>

Zimmi48 commented:
OK so FTR this file was introduced in a088d03434417e935df3c75f81a954eadbdfc2b8 and it seems like it isn't used anymore.

Zimmi48 commented:
So I would say @zapashcanon just leave this file untouched in your PR (and therefore it will be mergeable since everyone else agreed to the changes) then we'll remove this file in a separate PR.

Zimmi48 commented:
BTW if you do this update, rebasing your PR on master at the same time would ensure that CI passes fully (right now SF appears as broken but this was fixed in master).

SkySkimmer commented:
@MSoegtropIMC I think this is ready to merge, do you want to do it?

MSoegtropIMC commented:
@SkySkimmer : I am using cygwin and the merge scripts are not tested on cygwin yet, For this reason I would prefer to do my first merge after 8.8.

Weather is too nice to mess up your weekend ...

SkySkimmer commented:
If it fails couldn't you just paste the error in a new issue and have someone else merge? That doesn't seem like a risk to a weekend.
If you disagree I'll just merge it though, no need to make this PR wait another week+.

MSoegtropIMC commented:
Fine with me - I will go ahead.

I guess it is ok that tests didn't pass, since these are expected to pass after the merge ...

Zimmi48 commented:
Yes, these two tests can be ignored as unrelated failures.

MSoegtropIMC commented:
@Zimmi48 : I have a few more questions on the merge process:

1.) one file was not reviewed, but it was discussed above that the file is not used. Not sure if it is OK to merge or if this file should be removed from the commit.

2.) CI did not pass, but it was discussed above that this is due to an unrelated issue already fixed on master.

So should I

a) just merge as is since all issues have been discussed by the reviewers
b) create a branch, fix the stuff and merge
c) request the author to fix it and run CI again

In case I should create a branch, I am unsure if I should create it on my fork (MsoegtropIMC/coq) or on the original Coq repo (coq/coq). I would find it more natural to create it on the original Coq repo (otherwise I have to do yet another pull request ...) but I don't see any branches on the original Coq repo except the master and version branches so it looks like it is not very common to do this. I guess in either case the reviewer approval needs to be redone?

And one more note on https://github.com/coq/coq/blob/master/dev/doc/MERGING.md:

> git push upstream

only works on a fork (otherwise there is no upstream), but

> dev/tools/merge-pr

has an explicit check that the primary remote is the original Coq repo. This is a bit confusing. I would think running this on a clone of the original Coq repo is preferable (otherwise one probably should sync with upstream first to get the latest version of the merge script).

SkySkimmer commented:
> b) create a branch, fix the stuff and merge

Don't create a branch, just push to @zapashcanon's repository.

Zimmi48 commented:
I would say, just merge as-is. See #7180 for the removal of the dead script. I don't care if I have to rebase this PR after the merge of the present one.

MSoegtropIMC commented:
@SkySkimmer : ah ok - I didn't imagine that I have write rights on zapashcanon/coq. Is it so that write rights are generally inherited in forks on github?

Zimmi48 commented:
About the `upstream` business you have to remember that git is a distributed VCS. So your local clone isn't attached to your fork more than it is to the official (upstream) Coq repository.

For instance, the remotes of my local clone are the following:

```
$ git remote -v
SkySkimmer	https://github.com/SkySkimmer/coq (fetch)
SkySkimmer	https://github.com/SkySkimmer/coq (push)
herbelin	https://github.com/herbelin/github-coq (fetch)
herbelin	https://github.com/herbelin/github-coq (push)
mattam	https://github.com/mattam82/coq (fetch)
mattam	https://github.com/mattam82/coq (push)
origin	git@github.com:Zimmi48/coq.git (fetch)
origin	git@github.com:Zimmi48/coq.git (push)
ppedrot	https://github.com/ppedrot/coq (fetch)
ppedrot	https://github.com/ppedrot/coq (push)
upstream	git@github.com:coq/coq (fetch)
upstream	git@github.com:coq/coq (push)
```

As you can see there is in particular, `origin` which points to my fork and `upstream` which points to the upstream repository. But you may use different names on your machine, so just adapt the `git push upstream` command.

SkySkimmer commented:
When making a PR there is an opt-out option to let maintainers of the 
target repository push to the PR branch.
Being opt-out it is usually enabled.

Gaëtan Gilbert

On 05/04/2018 18:08, MSoegtropIMC wrote:
> @SkySkimmer <https://github.com/SkySkimmer> : ah ok - I didn't imagine 
> that I have write rights on zapashcanon/coq. Is it so that write rights 
> are generally inherited in forks on github?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coq/coq/pull/6809#issuecomment-378989719>, or mute 
> the thread 
> <https://github.com/notifications/unsubscribe-auth/ACWQ7CRawA4u94SJM2ygRSoLRUb656bnks5tlkGEgaJpZM4SOa_2>.
>

Zimmi48 commented:
To illustrate @SkySkimmer's explanation, here are some screenshots.

What you see when you open a new PR:

![screenshot-2018-4-5 coq coq](https://user-images.githubusercontent.com/1108325/38378058-da3d87f0-38fc-11e8-961d-2bfd9d907c32.png)

What you can see at the bottom of the present PR, confirming that the box wasn't unchecked:

![screenshot-2018-4-5 improve shell scripts by zapashcanon pull request 6809 coq coq](https://user-images.githubusercontent.com/1108325/38378088-f1d476f8-38fc-11e8-96a6-42281158d68a.png)

Zimmi48 commented:
That being said, I don't think that any action other than just merging as-is is warranted in this case.

zapashcanon commented:
I can remove the dead file and rebase against master to check if the CI pass if you want.

MSoegtropIMC commented:
@zapashcanon : if you can do it today, I would prefer it. But I can also merge right away.

zapashcanon commented:
Done.

Zimmi48 commented:
Sorry to bug you @zapashcanon but the removal of the file is not what I suggested. I don't like having this removal done in this commit as it doesn't make sense semantically. What I suggested was to undo the changes to this file, leave it untouched.

zapashcanon commented:
Fixed, had to bypass the pre-commit hook which wanted to fix the whitespace in the file no matter what.

Zimmi48 commented:
Thank you for your reactivity!

> had to bypass the pre-commit hook which wanted to fix the whitespace in the file no matter what

Interesting... What command did you use? @SkySkimmer might be interested in reproducing and debugging this...

SkySkimmer commented:
I imagine it was git commit --no-verify


Gaëtan Gilbert

On 05/04/2018 23:27, Théo Zimmermann wrote:
> Thank you for your reactivity!
> 
>     had to bypass the pre-commit hook which wanted to fix the whitespace
>     in the file no matter what
> 
> Interesting... What command did you use? @SkySkimmer 
> <https://github.com/SkySkimmer> might be interested in reproducing and 
> debugging this...
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coq/coq/pull/6809#issuecomment-379082138>, or mute 
> the thread 
> <https://github.com/notifications/unsubscribe-auth/ACWQ7Mv1hH5k7icSde-jSxRX91SFypPMks5tlowlgaJpZM4SOa_2>.
>

zapashcanon commented:
Indeed.

MSoegtropIMC commented:
@zapashcanon : thanks for running the extra mile! I will do a final review and then run the merge.

Btw.: I don't find it that convenient that you merge your commits into one (with --amend). It makes it hard to review the incremental changes. One doesn't want to have that many commits either. I guess two (the initially suggested version and the delta to the final one) would be most efficient.

MSoegtropIMC commented:
@Zimmi48 : The merge-pr script went through, but the push failed:

remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: Waiting on code owner review from Zimmi48.

Looks like you must also give your ok ...

But it looks way like a bug in github. When I do "Show all reviewers" it doesn't list you, but in the top line of the "Code owner review required" area, it says it needs your review. As far as I remember this came only after this final fixes.

I checked the file "relatify_with-respect-to_.sh" it is there again and the changes in it are undone. The review request for @ppedrot didn't go away, though.

Zimmi48 commented:
Sorry about that @MSoegtropIMC. It looks like an admin of the Coq repository accidentally activated the "code owner review required" setting. It should be good now.

MSoegtropIMC commented:
@Zimmi48 : thanks! It did work now (or at least I hope so ;-)
